### DISCUSSION TITLE
How to use async/await in client components?

### DISCUSSION THREAD

**Discussion by tofuwu-dev** [vercel/next.js#55211] [Help] [answered]
### Summary

Hi everyone,

I am having some troubles to use async/await functions in client components.
Regardug data fetching, I used swr dependency to get data via external api.
But in my other case, I dont need to fetch data via api but I need to write code that requires async/await while mapping data.

If I cant use async/await in client components, please kindly let me know how to implement such an function in client components.

Hope your kind answers!

**Comment by icyJoseph** [comment]
Hi,

Client Components can't be async, and hence, in the main body of a Client Component, you cannot use await, nor can you easily couple with Promises.

In general, what you want to do here, is couple with a promise. You couple with a promise by using a `useState` + `useEffect` combo.

What kind of asynchronous work are you planning to do in the client component? There's limitations, such as not having access to server side APIs and the creation of a client bundle, which is perhaps something you don't want to do.

You could map your data in a Server Component, and pass it as props as a Client Component.

If this was not enough, maybe you could explain a bit more, and I can help further.

**Reply by tofuwu-dev** [reply]
Thanks for your reply, @icyJoseph !

I just wanted to make a function called `getBalance` to get/show balance of tokens of my wallet when it's connected and it should be implemented with async/await.
Here are some code snippets and it's in Server Component.
```bash
// function.ts
export const getBalance = (token) => {
...
   const response = await client.queryContractSmart(token.address, queryMsg); // To get balance from token address
   return response; // Suppose it's balance of token.
...
}
```
```bash
// balance.tsx
"use client"
...
   <div>
        {
             tokenList.map((token) => (
                 <div>
                          <p>{getBalance(token)}</p>
                 </div>
        }
   </div>
...
```

So I would love to show balance of tokens from their address while mapping token list array.

**Reply by icyJoseph** [reply]
Right in this case, you can use first, a server component, preferably the page level component, which does the async/await call to getBalance.

```tsx
import { BalanceList } from './balance';

// page.tsx
export default async function Page() {
   const balance = await getBalance(token);
   return <BalanceList balance={balance} />
}
```

Now, where does the `token` parameter to `getBalance` come from?

**Reply by tofuwu-dev** [reply]
`token` param would be from `tokenList` - you can just think that `tokenList` is static data
So I get `token address` and pass the param to `getBalance` and calculate balance.

![image](https://github.com/vercel/next.js/assets/98631674/e77f4ac3-5dc5-4709-88f4-8b45afec11c7)

**Reply by icyJoseph** [reply]
Right, I had missied that `<p>` tag wrapping the `getBalance` call. 

Here the issue would be that you have a `use client` directive, and everything under this tree will be a client component.

So you can't do this as you have it. Also you'd need to map over the tokenList getting promises first. 

If `tokenList` is a statically known value:

```tsx
import { BalanceList } from './balance';
// I guess, import getBalance too

export default async function Page(){
    const balanceList = await Promise.all(tokenList.map(getBalance))

    return <BalanceList balanceList={balanceList} tokenList={tokenList} />
}
```

In this case, the tokenData is awaited on a server component, the page entry point to be precise, and then it calculates the balances list, and passes it to the client component.

**Reply by tofuwu-dev** [reply]
@icyJoseph yes right!
But I have some more logics except of getting balance of tokens and I have to use  `use client` directive to activate `useState` and `useEffect`.
And `tokenList` would be called in many other components and I get this `tokenList` from external API and I want to use this array in other children components via `context`.
```bash
...
   const tokenList = useTokenListInfo() // it's `tokenList` from context. so I should use `use client` directive right?
   const balanceList = await Promise.all(tokenList.map(getBalance)) // instead of static token list
...
```
In this case, what can I do? and in my case, If possible, I would love to add `each balance` based on `token list` so I will have tokenList including not only address but balance after this process.
Then, I can show each balance of tokens while mapping `tokenList` or I can also show only one balance of selected token, I think.

Sorry for my long text.

**Reply by icyJoseph** [reply]
In this case you either follow this pattern and always use a server component that wraps a client component. Which can be hard to stick to, if the server component gets somehow included into a client tree. You can do slots/children props to have a server tree that uses client components here and there:

- https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#interleaving-server-and-client-components look at the supported pattern

Or, use an effect to fetch the data using Promise.all and when that resolves set it to a React state, that's what I meant with a useState + useEffect combo, which is the traditional way to couple to asynchronous work inside React components.

**Reply by tofuwu-dev** [reply]
I thoguht about that pattern as well. but it doesnt match in my case.
If I use that pattern, I should make server component which shows balance list and the server component should be children of client component.

But for the balance of random selected token, should I make one more server component except of component shows balance list based on token list?

So I wanted to use `tokenList` array value instead of component.

And also there might be more cases, ofc. 
Even tho balance is the same, it's style might different slightly.
When we suppose if it is children of more nested components, it might be more complex.
Should I use `hoc` for that?

**Reply by tofuwu-dev** [reply]
@icyJoseph I could finally implement what I want to do.
I think I misunderstood your last message. Thanks for your guide!
I will mark this as answer!

**Reply by tofuwu-dev** [reply]
But I have one thing which I want to confirm.
Can I use async await functions in client componetns tho?

```bash
// client-component.tsx

"use client"
import React, {useEffect} from "react

export default function ClientComponent() {
  useEffect(() => {
    (async() => {
       await something();
    })()
  }, [])

  const something = async() => {
   // bla bla bla....
  }
}
```

**Reply by icyJoseph** [reply, marked as answer]
Yeah, as I said earlier, that's the traditional way to couple a React Component to a promise. Using state (to store the result, in case there's such a thing), and an effect to run the async work.


```tsx
"use client"
import React, { useEffect } from "react";

export default function ClientComponent() {
  const something = async () => {
    // bla bla bla....
  };

  useEffect(() => {
    async function doWork() {
      const result = await something();
      // setState or do something with result
    }

    doWork();
  }, []);

  return;
}
```

Note that this will defer the work to the client side too. So the HTML served to the browser won't have this data.

**Reply by tofuwu-dev** [reply]
I think I should be able to pass data as props to activate server components for every page. :)

**Reply by icyJoseph** [reply]
No worries, there's a lot of new concepts and such in the App Router... Good luck!
